packages/core/src/util/txInspector.ts

```diff
@@ -17,47 +17,49 @@ type TxInspector<T extends Inspectors> = (tx: TxAlonzo) => {
   [k in keyof T]: ReturnType<T[k]>;
 };
 
-// Inspectors result types
+// Inspection result types
 export type SendReceiveValueInspection = Value;
 export type DelegationInspection = StakeDelegationCertificate[];
 export type StakeKeyRegistrationInspection = StakeAddressCertificate[];
 export type WithdrawalInspection = Lovelace;
 
+// Inspector types
+export type SendReceiveValueInspector = (ownAddresses: Address[]) => Inspector<SendReceiveValueInspection>;
+export type DelegationInspector = Inspector<DelegationInspection>;
+export type StakeKeyRegistrationInspector = Inspector<StakeKeyRegistrationInspection>;
+export type WithdrawalInspector = Inspector<WithdrawalInspection>;
+
 /**
  * Inspects a transaction for value (coins + assets) sent by the provided addresses.
  *
  * @param {Address[]} ownAddresses own wallet's addresses
  * @returns {Value} total value sent
  */
-export const valueSentInspector =
-  (ownAddresses: Address[]): Inspector<Value> =>
-  (tx: TxAlonzo): SendReceiveValueInspection => {
-    if (!isOutgoing(tx, ownAddresses)) return { coins: 0n };
-    const sentOutputs = tx.body.outputs.filter((out) => !isAddressWithin(ownAddresses)(out));
-    return coalesceValueQuantities(sentOutputs.map((output) => output.value));
-  };
+export const valueSentInspector: SendReceiveValueInspector = (ownAddresses: Address[]) => (tx: TxAlonzo) => {
+  if (!isOutgoing(tx, ownAddresses)) return { coins: 0n };
+  const sentOutputs = tx.body.outputs.filter((out) => !isAddressWithin(ownAddresses)(out));
+  return coalesceValueQuantities(sentOutputs.map((output) => output.value));
+};
 
 /**
  * Inspects a transaction for value (coins + assets) received by the provided addresses.
  *
  * @param {Address[]} ownAddresses own wallet's addresses
  * @returns {Value} total value received
  */
-export const valueReceivedInspector =
-  (ownAddresses: Address[]): Inspector<Value> =>
-  (tx: TxAlonzo): SendReceiveValueInspection => {
-    if (isOutgoing(tx, ownAddresses)) return { coins: 0n };
-    const receivedOutputs = tx.body.outputs.filter((out) => isAddressWithin(ownAddresses)(out));
-    return coalesceValueQuantities(receivedOutputs.map((output) => output.value));
-  };
+export const valueReceivedInspector: SendReceiveValueInspector = (ownAddresses: Address[]) => (tx: TxAlonzo) => {
+  if (isOutgoing(tx, ownAddresses)) return { coins: 0n };
```

Comment:
Not a change in this PR, but this line is not entirely correct. You could have a tx that both sends and receives:
- from plutus script address
- with a tx signed by multiple parties (yourself and someone else)

Note: if we change this then TransactionsTracker.incoming$ will also be affected and we could have transactions that are both incoming and outgoing. Not sure if that would break anything :thinking: 

@rhyslbw @james-iohk thoughts?

Reply:
Right, this catch should be lower in the logic stack, as a fall-back value instread

Reply:
I see, but in that case how do you know which are change utxo and what are you actually receiving?

I could rename `isOutgoing` to `isSending` where "is sending" means "my addresses are included in the tx inputs" (I guess we could also rename it to `isWithinInputs` or something like that)

* `valueSentInspector`:

```ts
export const valueSentInspector: SendReceiveValueInspector = (ownAddresses: Address[]) => (tx: TxAlonzo) => {
  if (!isSending(tx, ownAddresses)) return { coins: 0n };
  // Filter out change utxo
  const sentOutputs = tx.body.outputs.filter((out) => !isAddressWithin(ownAddresses)(out));
  return coalesceValueQuantities(sentOutputs.map((output) => output.value));
};
```
If I'm not wrong that check in the `valueSentInspector` should still be fine, if your address is not in the tx inputs then you are not sending anything. And if it is included in some input, then all outputs that don't include your address is what you are actually sending.

But what about this case: 
> with a tx signed by multiple parties (yourself and someone else)

could someone else be sending something in that same tx? Could there be change utxo for that other wallet?

For example, could something like this happen in the same transaction?
- `MyWallet` sends to `WalletA` -> here you would have sending outputs to `WalletA` and change outputs for `MyWallet`  
- `WalletA` sends to `WalletB` -> here you would have sending outputs to `WalletB` and change outputs for `WalletA`

How do you differentiate between `sending outputs to WalletA` and `change outputs for WalletA`? Do we even care about that or should we just filter out `change outputs for MyWallet` and consider everything else as sending as we are doing now?

Reply:
- `valueReceivedInspector`:

```ts
export const valueReceivedInspector: SendReceiveValueInspector = (ownAddresses: Address[]) => (tx: TxAlonzo) => {
  if (isSending(tx, ownAddresses)) return { coins: 0n };
  const receivedOutputs = tx.body.outputs.filter((out) => isAddressWithin(ownAddresses)(out));
  return coalesceValueQuantities(receivedOutputs.map((output) => output.value));
};
```

For this function, as you mentioned, the `isSending` shouldn't be here. But could something like this happen?

- `MyWallet` sends to `WalletA` -> here you would have sending outputs to `WalletA` and change outputs for `MyWallet`
- `WalletA` sends to `MyWallet` -> here you would have sending outputs to `MyWallet` and change outputs for `WalletA`

In that case, we should only consider `sending outputs to MyWallet` from `WalletA` as the received value, which `tx.body.outputs.filter((out) => isAddressWithin(ownAddresses)(out))` would get, but it would also include `change outputs for MyWallet`, how do we identify them and filter them out?

Reply:
> But could something like this happen

I guess it's technically possible but would be unpractical.

I'm now thinking that the concept of incoming$ and outgoing$ observables is just too abstract and difficult to define. Maybe we should just leave it up to the SDK user to apply the filters he wants. How about refactoring the interface to have just `transactions.history$` (the current `all$`), intended to be filtered with the inspectors?

If we do that then we should definitely remove logic like this: `if (isSending(tx, ownAddresses)) return { coins: 0n };` and have well named inspectors for:
1. sum value of all inputs from own addr
2. sum value of all outputs to own addr
3. net value spent = 1 - 2
4. net value received = 2 - 1

Reply:
Ok, I could take a look at this. Should I do it in this PR or should we merge this one as it is and create a new branch for this changes?

Some questions:
> 1. sum value of all inputs from own addr
Inputs do not have value, how can I get this without using the wallet provider?

> 5. could also do a "did I have to sign the transaction" inspector that would check 1. + stuff in certificates. I think that's the typical "send transaction" classification that a user might expect
I don't think I understand this one

Reply:
> I'm now thinking that the concept of incoming$ and outgoing$ observables is just too abstract and difficult to define. Maybe we should just leave it up to the SDK user to apply the filters he wants. How about refactoring the interface to have just transactions.history$ (the current all$), intended to be filtered with the inspectors?

About this, we can remove `history.incoming$` and `history.outgoing$`, but what about the others:
```ts
history: {
  all$: historicalTransactions$,
  incoming$: incomingTransactionHistory$,
  outgoing$: outgoingTransactionHistory$
},
incoming$: newTransactions$(incomingTransactionHistory$),
outgoing: {
  confirmed$,
  failed$,
  inFlight$,
  pending$,
  submitting$
},
```
Most depend on `outgoingTransactionHistory$` and `incomingTransactionHistory$` in one way or another.
The light wallet only uses `outgoing.inFlight$` though, which needs `txConfirmed$` and that one needs `outgoingTransactionHistory$` which I think could be replaced by `historicalTransactions$`

Should we maybe leave it like this?
```ts
{
  history$: historicalTransactions$,
  confirmed$,
  failed$,
  inFlight$,
  pending$,
  submitting$
},
```
after making `txConfirmed$` use `historicalTransactions$` instead of `outgoingTransactionHistory$` if it makes sense

Reply:
> Should I do it in this PR or should we merge this one as it is and create a new branch for these changes?

I think merging this is better, but let's do that tomorrow, so if you can get some progress in the mean time then great.

> Inputs do not have value, how can I get this without using the wallet provider?

Right, seems like this will require resolving it with the provider. We should probably change the return type of transactions fetched via provider. Let's scope this out of this PR.

> I don't think I understand this one

I think most users would consider any transactions that they sign with their wallet as "sent", so we probably want to have an inspector for that. My understanding is that a way to recognize such transaction is either:

- it has any inputs from own addr
- it has any certificates referencing own reward account. I'm actually not 100% sure about the rules of signature validation for all certificate types, but an example would be a delegation certificate - you need to sign such transactions with your stake key

It is unlikely that a tx would have such certificates without any inputs from own address, but I think it's technically possible - tx fee could be paid with an input owned by plutus script address for example, or just a different wallet.

> Should we maybe leave it like this?

Sorry if I wasn't clear on my suggestion here. I wanted to suggest that we should probably replace this:

```
history: {
  all$: historicalTransactions$,
  incoming$: incomingTransactionHistory$,
  outgoing$: outgoingTransactionHistory$
},
```
with `history$: historicalTransactions$,`